### PR TITLE
Mod filesystem

### PR DESCRIPTION
**Pull request by Idhrendur**

### CODE REVIEW

File: ModLoader/ModFilesystem.cpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+	{
+		for (const auto& new_file: commonItems::GetAllFilesInFolder(mod.path + '/' + std::string(path)))
+		{
+			if (found_files.contains(new_file))
+			{
+				continue;
+			}
+
+			found_files.insert(new_file);
+			full_files.insert(mod.path + '/' + path + '/' + std::string(new_file));
```
Comment (Zemurin) [inline ModLoader/ModFilesystem.cpp:95]: Hmm. I believe you're inserting the first match in a mod that contains this file and ignoring further instances. However if a second mod is a compatch that has current mod as a dependency then it should override this file with corrected one.

File: ModLoader/ModFilesystem.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+	for (const auto& new_file: commonItems::GetAllFilesInFolder(game_root_ + '/' + std::string(path)))
+	{
+		if (found_files.contains(new_file))
+		{
+			continue;
+		}
+
+		found_files.insert(new_file);
+		full_files.insert(game_root_ + '/' + path + '/' + std::string(new_file));
```
Comment (Zemurin) [inline ModLoader/ModFilesystem.cpp:112]: Ditto here.

File: ModLoader/ModFilesystem.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+	return full_files;
+}
+
+
+std::set<std::string> commonItems::ModFilesystem::GetAllSubfolders(const std::string& path) const
+{
+	std::set<std::string> full_folders;
+	std::set<std::string> found_folders;
+
+	for (const auto& mod: mods_ | std::views::reverse)
```
Comment (Zemurin) [inline ModLoader/ModFilesystem.cpp:124]: In fact, if we didn't use a plain vector iteration here for mods, but somehow ordered the mods beforehand according to dependencies then this code should work fine - if the highest-dependency mods were on the end of the vector and you started at the end. I see you're reversing order you're loading the mods from but the original order is alphabetical, not dependency-based, no? Straight scoop from disk.

This also related to the Latest*Determines* Tests which shouldn't be the case.

**Review by Zemurin — CHANGES_REQUESTED** [review]
Aside for the ordering logic hitch, rest looks perfectly fine.

File: ModLoader/ModFilesystem.cpp [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+
+
+namespace
+{
+
+bool PathIsReplaced(std::string_view path, const std::set<std::string>& replaced_paths)
+{
+	return std::ranges::any_of(replaced_paths, [&path](const std::string& replaced_path) {
+		if (path.starts_with(replaced_path))
```
Comment (IhateTrains) [inline ModLoader/ModFilesystem.cpp:17]: What if `path` is a substring of `replaced_path`, but actually a different folder, e.g.
```
path = common/culture_groups
replacedpath = common/culture
```
?

**Review by IhateTrains — APPROVED** [review]

File: ModLoader/ModFilesystem.cpp (hunk 1 above)
Comment (IhateTrains) [reply ModLoader/ModFilesystem.cpp:95]: Do we really need to handle compatches in a special way? A compatch should be last in the load order, so if we load the mods in load order and iterate over that but reversed, we will get the overridden file from compatch mod, no?

Comment (Zemurin) [reply ModLoader/ModFilesystem.cpp:95]: No, that't not the case, unfortunately. Apparently, at least for older games like EU4 and HoI4, the load order provided by the launcher does absolutely nothing, the game sorts the load order internally according to dependencies listed in the .mod file.

But we don't even see the mod load order, we grab all the files from disk alphabetically, and with ugc_<crap>.mod filenames mod order is in fact completely random.

Comment (IhateTrains) [reply ModLoader/ModFilesystem.cpp:95]: WTF why would they do that. I don't want to touch a non-Jomini game ever again.

Comment (IhateTrains) [reply ModLoader/ModFilesystem.cpp:95]: I guess we could add an `is_jomini` parameter to `ModFilesystem` constructor, and depending on that use the load order logic or the older game dependencies "logic".

Comment (IhateTrains) [reply ModLoader/ModFilesystem.cpp:95]: Or just drop older games, EU5 is probably coming relatively soon 😈

Comment (Idhrendur) [reply ModLoader/ModFilesystem.cpp:95]: The intent was that the caller passes in the mods in the intended order. So it's up to the caller to sort based on dependencies. I'll add a function to the mod loader in another PR after merging this.

File: ModLoader/ModFilesystem.cpp (hunk 3 above)
Comment (Idhrendur) [reply ModLoader/ModFilesystem.cpp:124]: As mentioned on the other comment, the caller of this library determines the order. I've added a clarifying comment in the header.

File: ModLoader/ModFilesystem.cpp (hunk 4 above)
Comment (Idhrendur) [reply ModLoader/ModFilesystem.cpp:17]: Oh, shoot.

Comment (Idhrendur) [reply ModLoader/ModFilesystem.cpp]: Fixed

**Review by Zemurin — APPROVED** [review]

**Pull request merged**